MacrobenchmarkSample/macrobenchmark/src/main/java/com/example/macrobenchmark/startup/StartupCompilationModeBenchmark.kt

```diff
@@ -0,0 +1,72 @@
+package com.example.macrobenchmark.startup
+
+import androidx.benchmark.macro.BaselineProfileMode
+import androidx.benchmark.macro.CompilationMode
+import androidx.benchmark.macro.StartupMode
+import androidx.benchmark.macro.StartupTimingMetric
+import androidx.benchmark.macro.junit4.MacrobenchmarkRule
+import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+/**
+ * Run this benchmark from Studio to see startup measurements, and captured system traces
+ * for investigating your app's performance from a cold state.
+ */
+@RunWith(AndroidJUnit4ClassRunner::class)
+class ColdStartupBenchmark : AbstractStartupBenchmark(StartupMode.COLD)
+
+/**
+ * Run this benchmark from Studio to see startup measurements, and captured system traces
+ * for investigating your app's performance from a warm state.
+ */
+@RunWith(AndroidJUnit4ClassRunner::class)
+class WarmStartupBenchmark : AbstractStartupBenchmark(StartupMode.WARM)
+
+/**
+ * Run this benchmark from Studio to see startup measurements, and captured system traces
+ * for investigating your app's performance from a hot state.
+ */
+@RunWith(AndroidJUnit4ClassRunner::class)
+class HotStartupBenchmark : AbstractStartupBenchmark(StartupMode.HOT)
+
+/**
+ * Base class for benchmarks with different startup modes.
+ * Enables app startups from various states of baseline profile or [CompilationMode]s.
+ */
+abstract class AbstractStartupBenchmark(private val startupMode: StartupMode) {
+    @get:Rule
+    val benchmarkRule = MacrobenchmarkRule()
+
+    @Test
+    fun startupNoCompilation() = startup(CompilationMode.None())
+
+    @Test
+    fun startupBaselineProfileDisabled() = startup(
+        CompilationMode.Partial(
+            baselineProfileMode = BaselineProfileMode.Disable,
+            warmupIterations = 1
+        )
+    )
+
+    @Test
+    fun startupBaselineProfile() =
+        startup(CompilationMode.Partial(baselineProfileMode = BaselineProfileMode.Require))
+
+    @Test
+    fun startupFullCompilation() = startup(CompilationMode.Full())
+
+    private fun startup(compilationMode: CompilationMode) = benchmarkRule.measureRepeated(
+        packageName = "com.example.macrobenchmark.target",
+        metrics = listOf(StartupTimingMetric()),
+        compilationMode = compilationMode,
+        iterations = 10,
+        startupMode = startupMode,
+        setupBlock = {
+            pressHome()
```

Comment:
Curious, why is pressHome needed.  It's not in the doc sample.

https://developer.android.com/studio/profile/macrobenchmark-overview

Reply:
It's part of the [MacrobenchmarkRule](https://developer.android.com/reference/kotlin/androidx/benchmark/macro/junit4/MacrobenchmarkRule) docs (where it'll be corrected to use `startupBlock` soon.
It ensures the benchmark is started from the home screen in any case.

Reply:
OK, but should we recommend most developers include this? Should I include it?

Reply:
`pressHome()` is necessary if you are doing a warm or a hot start. For cold we kill the process anyway.

Reply:
Would you consider putting pressHome() in the default implementation of startupBlock? Particularly when you change the name it's the ideal change to simplify this.

I'm unclear from the docs what happens when startupMode isn't specified.  It seems distinct from any of the other modes.

Docs link to https://github.com/android/performance-samples/blob/b576634530e6069dfb1e8ec4003cd78f521dc5a2/MacrobenchmarkSample/macrobenchmark/src/main/java/com/example/macrobenchmark/startup/SampleStartupBenchmark.kt

Reply:
@mlykotom can you fix that docs link/sample? every startup benchmark should call "pressHome()" first.

> Would you consider putting pressHome() in the default implementation of startupBlock? Particularly when you change the name it's the ideal change to simplify this.

What name change are you referring to? setupBlock doesn't assume the user is writing a startup benchmark - it's valid for someone to launch the app exactly once (first setup block) and never press home. having default behavior in pressHome may make that confusing. (See 
https://developer.android.com/reference/kotlin/androidx/benchmark/macro/junit4/MacrobenchmarkRule#measurerepeated which describes exact looping behavior)

In an ideal world, `startupMode` would be an argument to `startActivityAndWait`, since it's a property of that launch. In reality though, a hot or warm start means the process is already alive, and potentially already showing a given activity instance. We could start the activity twice, but we're already tracing within measureBlock, so we'd have to be careful to not measure that startup (within a measureBlock, which is counter intuitive).

Instead, startupMode is test-wide configuration that 1) defines the process state before measurement is started and 2) affects whether activityStartAndWait uses `Intent.FLAG_ACTIVITY_CLEAR_TASK` (effectively, set to prevent reusing activities when startupMode != hot).

Reply:
Rahul gave me more context here, that he thought it was improving stability. Filed https://issuetracker.google.com/issues/226659872 to test it out in androidx benchmarks.